From: Takashi Iwai <tiwai@suse.de>
Subject: [ALSA] pcm-oss - Add bugg-yptr option
Patch-mainline: 
References: 115829

Added "buggy-ptr" proc option to switch the behavior of GETOPTR ioctl.

Signed-off-by: Takashi Iwai <tiwai@suse.de>

---


diff -u linux/Documentation/sound/alsa/ALSA-Configuration.txt:1.87 linux/Documentation/sound/alsa/ALSA-Configuration.txt:1.88
--- linux/Documentation/sound/alsa/ALSA-Configuration.txt:1.87	Mon Sep  5 09:12:29 2005
+++ linux/Documentation/sound/alsa/ALSA-Configuration.txt	Thu Sep  8 05:48:34 2005
@@ -1592,6 +1592,8 @@
 	  - whole-frag  write only whole fragments (optimization affecting
 			playback only)
 	  - no-silence  do not fill silence ahead to avoid clicks
+	  - buggy-ptr	Returns the whitespace blocks in GETOPTR ioctl
+			instead of filled blocks
 
   Example: echo "x11amp 128 16384" > /proc/asound/card0/pcm0p/oss
            echo "squake 0 0 disable" > /proc/asound/card0/pcm0c/oss
diff -u linux/sound/core/oss/pcm_oss.c:1.97 linux/sound/core/oss/pcm_oss.c:1.98
--- linux/sound/core/oss/pcm_oss.c:1.97	Mon Sep  5 02:35:21 2005
+++ linux/sound/core/oss/pcm_oss.c	Thu Sep  8 05:48:34 2005
@@ -1543,7 +1543,11 @@
 	} else {
 		delay = snd_pcm_oss_bytes(substream, delay);
 		if (stream == SNDRV_PCM_STREAM_PLAYBACK) {
-			info.blocks = (runtime->oss.buffer_bytes - delay - fixup) / runtime->oss.period_bytes;
+			snd_pcm_oss_setup_t *setup = substream->oss.setup;
+			if (setup && setup->buggyptr)
+				info.blocks = (runtime->oss.buffer_bytes - delay - fixup) / runtime->oss.period_bytes;
+			else
+				info.blocks = (delay + fixup) / runtime->oss.period_bytes;
 			info.bytes = (runtime->oss.bytes - delay) & INT_MAX;
 		} else {
 			delay += fixup;
@@ -2350,6 +2354,8 @@
 				template.partialfrag = 1;
 			} else if (!strcmp(str, "no-silence")) {
 				template.nosilence = 1;
+			} else if (!strcmp(str, "buggy-ptr")) {
+				template.buggyptr = 1;
 			}
 		} while (*str);
 		if (setup == NULL) {
diff -u linux/include/sound/pcm_oss.h:1.11 linux/include/sound/pcm_oss.h:1.12
--- linux/include/sound/pcm_oss.h:1.11	Fri Oct 22 23:56:32 2004
+++ linux/include/sound/pcm_oss.h	Thu Sep  8 05:48:34 2005
@@ -32,7 +32,8 @@
 		     block:1,
 		     nonblock:1,
 		     partialfrag:1,
-		     nosilence:1;
+		     nosilence:1,
+		     buggyptr:1;
 	unsigned int periods;
 	unsigned int period_size;
 	snd_pcm_oss_setup_t *next;
